internal/auth/internal/types/traffic_permissions.go

```diff
@@ -31,114 +31,239 @@ func RegisterTrafficPermissions(r resource.Registry) {
 		Proto:    &pbauth.TrafficPermissions{},
 		Scope:    resource.ScopeNamespace,
 		Validate: ValidateTrafficPermissions,
+		Mutate:   MutateTrafficPermissions,
 	})
 }
 
+func MutateTrafficPermissions(res *pbresource.Resource) error {
+	var tp pbauth.TrafficPermissions
+
+	if err := res.Data.UnmarshalTo(&tp); err != nil {
+		return resource.NewErrDataParse(&tp, err)
+	}
+
+	var changed bool
+
+	for _, p := range tp.Permissions {
+		for _, s := range p.Sources {
+			if updated := normalizedTenancyForSource(s, res.Id.Tenancy); updated {
+				changed = true
+			}
+		}
+	}
+
+	if !changed {
+		return nil
+	}
+
+	return res.Data.MarshalFrom(&tp)
+}
+
+func normalizedTenancyForSource(src *pbauth.Source, parentTenancy *pbresource.Tenancy) bool {
+	var changed bool
+
+	if t, c := defaultedSourceTenancy(src, parentTenancy); c {
+		src.Partition = t.Partition
+		src.Peer = t.PeerName
+		src.Namespace = t.Namespace
+		changed = true
+	}
+
+	for _, e := range src.Exclude {
+		if t, c := defaultedSourceTenancy(e, parentTenancy); c {
+			e.Partition = t.Partition
+			e.Peer = t.PeerName
+			e.Namespace = t.Namespace
+			changed = true
+		}
+	}
+
+	return changed
+}
+
+func defaultedSourceTenancy(s pbauth.SourceToSpiffe, parentTenancy *pbresource.Tenancy) (*pbresource.Tenancy, bool) {
+	if !isLocalPeer(s.GetPeer()) {
+		return nil, false
+	}
+
+	tenancy := pbauth.SourceToTenancy(s)
+
+	var peerChanged bool
+	tenancy.PeerName, peerChanged = firstNonEmptyString(tenancy.PeerName, parentTenancy.PeerName, resource.DefaultPeerName)
+
+	var partitionChanged bool
+	tenancy.Partition, partitionChanged = firstNonEmptyString(tenancy.Partition, parentTenancy.Partition, resource.DefaultPartitionName)
+
+	var namespaceChanged bool
+	if s.GetIdentityName() != "" && tenancy.Partition == parentTenancy.Partition {
```

Comment:
This wouldn't default the namespace if the partitions differ. For that you'd need:
```
if s.GetIdentityName() != "" {
    if tenancy.Partition == parentTenancy.Partition {
        // what you had
    } else {
        // what you had, except use resource.DefaultNamespaceName for both args 2 and 3, ignoring parentTenancy
    }
}
```